From fcd45762783eae8dc2b0498248c1c6dad0c03320 Mon Sep 17 00:00:00 2001
From: Srikanth Thokala <sthokal@xilinx.com>
Date: Wed, 13 Mar 2013 20:00:34 -0700
Subject: [PATCH 106/628] xilinx: vdma: External frame sync issue

git://github.com/Xilinx/linux-xlnx.git xilinx-v14.7
commit a7e5b975c2b45ef5b2180a21f9eed897550fdb9e

This fixes external frame sync issue.

Signed-off-by: SrikanthT <sthokal@xilinx.com>
Signed-off-by: Radhey Shyam Pandey <radheys@xilinx.com>
Signed-off-by: Liming Wang <liming.wang@windriver.com>
---
 drivers/dma/xilinx/xilinx_axivdma.c |    9 ++++++---
 1 files changed, 6 insertions(+), 3 deletions(-)

diff --git a/drivers/dma/xilinx/xilinx_axivdma.c b/drivers/dma/xilinx/xilinx_axivdma.c
index 180ed14..d7e2098 100644
--- a/drivers/dma/xilinx/xilinx_axivdma.c
+++ b/drivers/dma/xilinx/xilinx_axivdma.c
@@ -41,6 +41,8 @@
 						/* Reset DMA engine */
 #define XILINX_VDMA_CR_RUNSTOP_MASK	0x00000001
 						/* Start/stop DMA engine */
+#define XILINX_VDMA_CR_FSYNC_SRC_MASK	0x00000060
+						/* FSYNC Source Mask */
 
 #define XILINX_VDMA_SR_HALTED_MASK	0x00000001
 						/* DMA channel halted */
@@ -92,7 +94,7 @@
 #define XILINX_VDMA_FRMCNT_EN	0x00000010	/* Frm Cnt enable mode */
 #define XILINX_VDMA_MSTR_MASK	0x00000F00	/* Master in control */
 
-#define XILINX_VDMA_EXTFSYNC_SHIFT	6
+#define XILINX_VDMA_EXTFSYNC_SHIFT	5
 #define XILINX_VDMA_MSTR_SHIFT		8
 #define XILINX_VDMA_WR_REF_SHIFT	8
 
@@ -1009,8 +1011,9 @@ static int xilinx_vdma_device_control(struct dma_chan *dchan,
 
 		chan->config.disable_intr = cfg->disable_intr;
 
-		if (cfg->ext_fsync)
-			reg |= cfg->ext_fsync << XILINX_VDMA_EXTFSYNC_SHIFT;
+		/* FSync Source selection */
+		reg &= ~XILINX_VDMA_CR_FSYNC_SRC_MASK;
+		reg |= cfg->ext_fsync << XILINX_VDMA_EXTFSYNC_SHIFT;
 
 		VDMA_OUT(&chan->regs->cr, reg);
 		return 0;
-- 
1.7.5.4

